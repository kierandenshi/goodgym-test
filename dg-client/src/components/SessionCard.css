.card {
  margin: 1rem 0 1rem 0;
  padding: 1rem;
  border-radius: 5px;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
}

.card-title {
  margin-right: 0.5rem;
  padding: 0.25rem 0.5rem;
  background: grey;
  border-radius: 5px;
}

.card-avatars {
  display: flex;
}
