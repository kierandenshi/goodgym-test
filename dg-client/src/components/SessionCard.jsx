import React from 'react';
import PropTypes from 'prop-types';
import Avatar from './Avatar';
import './SessionCard.css';

export default function SessionCard({ session: { name, area, goodgymers }, currentUser }) {
  return (
    <div className='card'>
      <div className='card-header'>
        <span className='card-title'>{area.name}</span>
        <span className='card-title'>{name}</span>
      </div>
      <div className='card-body'>
        <p>
          Etiam porta sem malesuada magna mollis euismod. Aenean eu leo quam.
          Pellentesque ornare sem lacinia quam venenatis vestibulum. Lorem sociis
          natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
        </p>
        {goodgymers.length ? <>
          <div className='card-avatars'>
            {goodgymers.map(goodgymer => (
              <Avatar key={goodgymer.id} goodgymer={goodgymer} />
            ))}
          </div>
          <div>{attendanceString(goodgymers, currentUser)}</div>
        </> : null}
        {isAttending(goodgymers, currentUser) ? <button>Cancel</button> : <button>Register</button>}
      </div>
    </div>
  );
}

function isAttending(goodgymers, currentUser) {
  return goodgymers.find(goodgymer => goodgymer.id === currentUser);
}
function attendanceString(goodgymers, currentUser) {
  if (isAttending(goodgymers, currentUser)) {
    return goodgymers.length > 1 ? `You and ${goodgymers.length} other GoodGymers are going` : `You and 1 other GoodGymer is going`;
  }
  return goodgymers.length > 1 ? `${goodgymers.length} GoodGymers are going` : `1 GoodGymer is going`;
}

SessionCard.propTypes = {
  session: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    area: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    }).isRequired,
    goodgymers: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      first_name: PropTypes.string.isRequired,
      last_name: PropTypes.string.isRequired,
    }))
  }).isRequired
}
