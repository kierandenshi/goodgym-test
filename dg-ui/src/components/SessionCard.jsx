import PropTypes from 'prop-types';
import Avatar from './Avatar';
import './SessionCard.css';

export default function SessionCard({
  session: { id, name, description, area, goodgymers }, currentUser, joinSession, leaveSession
}) {
  return (
    <div className='card'>
      <div className='card-header'>
        <span className='card-title blue'>{area.name}</span>
        <span className='card-title'>{name}</span>
      </div>
      <div className='card-body'>
        <p>{description}</p>
        {goodgymers.length ? <>
          <div className='card-avatars'>
            {goodgymers.map(goodgymer => (
              <Avatar key={goodgymer.id} goodgymer={goodgymer} />
            ))}
          </div>
          <div>{attendanceString(goodgymers, currentUser)}</div>
        </> : null}
        {isAttending(goodgymers, currentUser) ? <button onClick={() => leaveSession(id)}>Leave session</button> : <button onClick={() => joinSession(id)}>Join session</button>}
      </div>
    </div>
  );
}

function isAttending(goodgymers, currentUser) {
  return goodgymers.find(goodgymer => goodgymer.id === currentUser);
}

function attendanceString(goodgymers, currentUser) {
  if (isAttending(goodgymers, currentUser)) {
    return goodgymers.length > 2 ? `You and ${goodgymers.length -1} other GoodGymers are going` : `You and 1 other GoodGymer is going`;
  }
  return goodgymers.length > 1 ? `${goodgymers.length} GoodGymers are going` : `1 GoodGymer is going`;
}

SessionCard.propTypes = {
  session: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    area: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    }).isRequired,
    goodgymers: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      first_name: PropTypes.string.isRequired,
      last_name: PropTypes.string.isRequired,
    }))
  }).isRequired,
  currentUser: PropTypes.string.isRequired,
  joinSession: PropTypes.func.isRequired,
  leaveSession: PropTypes.func.isRequired
}
