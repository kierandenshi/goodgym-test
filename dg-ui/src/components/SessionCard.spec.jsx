import { describe, expect, it } from "vitest";
import { render } from "@testing-library/react";
import SessionCard from "./SessionCard";

describe("UserSelect test", () => {
  const session = {
    id: "123",
    name: "Mega session",
    description: "This is a mega session",
    area: { id: "123", name: "Brighton & Hove" },
    goodgymers: [
      {
        id: "123",
        first_name: "Kilgore",
        last_name: "Trout",
      },
      {
        id: "124",
        first_name: "Billy",
        last_name: "Pilgrim",
      },
    ],
  };
  const currentUser = "123";
  const joinSession = () => null;
  const leaveSession = () => null;
  const component = render(
    <SessionCard
      session={session}
      currentUser={currentUser}
      joinSession={joinSession}
      leaveSession={leaveSession}
    />,
  );

  it("should match the component snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
