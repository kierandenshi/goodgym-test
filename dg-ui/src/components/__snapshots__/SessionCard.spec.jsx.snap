// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserSelect test > should match the component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header"
        >
          <span
            class="card-title"
          >
            Brighton & Hove
          </span>
          <span
            class="card-title"
          >
            Mega session
          </span>
        </div>
        <div
          class="card-body"
        >
          <p>
            This is a mega session
          </p>
          <div
            class="card-avatars"
          >
            <div
              class="avatar"
              title="Kilgore Trout"
            >
              <span>
                K
                T
              </span>
            </div>
            <div
              class="avatar"
              title="Billy Pilgrim"
            >
              <span>
                B
                P
              </span>
            </div>
          </div>
          <div>
            You and 1 other GoodGymer is going
          </div>
          <button>
            Leave session
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
    >
      <div
        class="card-header"
      >
        <span
          class="card-title"
        >
          Brighton & Hove
        </span>
        <span
          class="card-title"
        >
          Mega session
        </span>
      </div>
      <div
        class="card-body"
      >
        <p>
          This is a mega session
        </p>
        <div
          class="card-avatars"
        >
          <div
            class="avatar"
            title="Kilgore Trout"
          >
            <span>
              K
              T
            </span>
          </div>
          <div
            class="avatar"
            title="Billy Pilgrim"
          >
            <span>
              B
              P
            </span>
          </div>
        </div>
        <div>
          You and 1 other GoodGymer is going
        </div>
        <button>
          Leave session
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
